samples/nrf_rpc/protocols_serialization/client/src/bt_test_shell.c

```diff
@@ -0,0 +1,277 @@
+
+typedef struct {
+	uint8_t data[TEST_DATA_SIZE];
+	bool notify_enabled;
+	bool indicate_enabled;
+	struct bt_conn *conn;
+} test_svc_ctx_t;
+
+static test_svc_ctx_t test_ctx = {
+	.data = {},
+	.notify_enabled = false,
+	.indicate_enabled = false,
+	.conn = NULL,
+};
+
+BT_GATT_SERVICE_DEFINE(custom_svc, BT_GATT_PRIMARY_SERVICE(&custom_service_uuid),
+		       BT_GATT_CHARACTERISTIC(&custom_char_uuid.uuid,
+					      BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE |
+						      BT_GATT_CHRC_NOTIFY | BT_GATT_CHRC_INDICATE,
+					      BT_GATT_PERM_READ | BT_GATT_PERM_WRITE,
+					      read_custom_char, write_custom_char, &test_ctx),
+		       BT_GATT_CCC(ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE));
+
+static ssize_t read_custom_char(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+				uint16_t len, uint16_t offset)
+{
+	LOG_INF("BT Test Read. off %d len %d", offset, len);
+
+	if (offset >= TEST_DATA_SIZE) {
+		LOG_ERR("BT Test Invalid offset");
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	test_svc_ctx_t *ctx = (test_svc_ctx_t *)attr->user_data;
+	uint16_t actual_len = MIN(len, TEST_DATA_SIZE - offset);
+
+	LOG_HEXDUMP_INF(ctx->data, TEST_DATA_SIZE, "BT Test State of data");
+	LOG_HEXDUMP_INF(ctx->data + offset, actual_len, "BT Test Read:");
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, ctx->data, TEST_DATA_SIZE);
+}
+
+static ssize_t write_custom_char(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				 const void *buf, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	if (offset >= TEST_DATA_SIZE) {
+		LOG_ERR("BT Test Invalid offset");
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	test_svc_ctx_t *ctx = (test_svc_ctx_t *)attr->user_data;
+	uint16_t actual_len = MIN(len, TEST_DATA_SIZE - offset);
+
+	if (actual_len < len) {
+		LOG_WRN("BT Test Write truncated: requested %d, writing %d", len, actual_len);
+	}
+
+	memcpy(ctx->data + offset, buf, actual_len);
+	LOG_INF("BT Test write data at offset %d", offset);
+	LOG_HEXDUMP_INF(buf, actual_len, "BT Test write data");
+	LOG_HEXDUMP_INF(ctx->data, TEST_DATA_SIZE, "BT Test Final state of data after write");
+	update_subscribed(conn, ctx->data);
+	return actual_len;
+}
+
+static void ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	test_ctx.notify_enabled = (value & BT_GATT_CCC_NOTIFY) != 0;
+	test_ctx.indicate_enabled = (value & BT_GATT_CCC_INDICATE) != 0;
+	LOG_INF("BT Test Notify: %s, Indicate: %s", test_ctx.notify_enabled ? "ON" : "OFF",
+		test_ctx.indicate_enabled ? "ON" : "OFF");
+}
+
+static void indicate_cb(struct bt_conn *conn, struct bt_gatt_indicate_params *params, uint8_t err)
+{
+	if (err) {
+		LOG_ERR("BT Test Indication confirmation error: %d", err);
+		return;
+	}
+
+	LOG_INF("BT Test Indication confirmed by client");
+}
+
+static void update_subscribed(struct bt_conn *conn, void *param)
+{
+	uint8_t *data = (uint8_t *)param;
+	int err;
+
+	if (test_ctx.conn && test_ctx.notify_enabled) {
```

Comment:
It seems you don't really need `test_ctx.conn` as you always have `conn` argument, like in this case, right?